### PR TITLE
Serverside Pagination for payments

### PR DESCRIPTION
**Pull request by jackstar12**
This PR moves the pagination, sorting and searching of payments from the frontend to the backend, boosting efficiency.

Note that this is a breaking change to the payments api because the total amount of records has to be returned aswell for pagination to work.

Essential Additions:
- `fetch_page` utility to paginate a query.
- `FromRowModel` is a convenient class to declare a model can be parsed from a row (used by `fetch_page` to parse rows)
- `FilterModel` allows models to declare fields that can be searched by (included in openapi docs aswell)

### CODE REVIEW

File: lnbits/core/views/api.py [hunk 1]
```diff
@@ -138,7 +139,37 @@ async def api_payments(
...
+
+@core_app.get(
+    "/api/v1/payments/paginated",
+    name="Payment List",
+    summary="get paginated list of payments",
+    response_description="list of payments",
+    response_model=Page[Payment],
+    openapi_extra=generate_filter_params_openapi(Payment),
+)
+async def api_payments(
```
Comment (dni) [inline lnbits/core/views/api.py]: def api_payments -> defined twice

File: lnbits/core/crud.py [hunk 2]
```diff
@@ -404,6 +404,47 @@ async def get_payments(
...
+
+    if not filters:
+        filters = Filters()
+
+    if limit:
+        filters.limit = limit
+    if offset:
+        filters.offset = offset
+
+    page = await get_payments_paginated(
```
Comment (dni) [inline lnbits/core/crud.py:433]: do we want to  return paginated here aswell?

File: lnbits/core/views/api.py [hunk 3]
```diff
@@ -277,8 +308,8 @@ async def api_payments_pay_invoice(bolt11: str, wallet: Wallet):
     status_code=HTTPStatus.CREATED,
 )
 async def api_payments_create(
-    wallet: WalletTypeInfo = Depends(require_invoice_key),
-    invoiceData: CreateInvoiceData = Body(...),
+        wallet: WalletTypeInfo = Depends(require_invoice_key),
+        invoiceData: CreateInvoiceData = Body(...),
```
Comment (dni) [inline lnbits/core/views/api.py]: tab to much?

File: lnbits/core/views/api.py [hunk 4]
```diff
@@ -309,7 +340,7 @@ class CreateLNURLData(BaseModel):
 
 @core_app.post("/api/v1/payments/lnurl")
 async def api_payments_pay_lnurl(
-    data: CreateLNURLData, wallet: WalletTypeInfo = Depends(require_admin_key)
+        data: CreateLNURLData, wallet: WalletTypeInfo = Depends(require_admin_key)
```
Comment (dni) [inline lnbits/core/views/api.py]: tab to much?

**Comment by jackstar12** [comment]
> > Note that this is a breaking change to the payments api because the total amount of records has to be returned aswell for pagination to work.
> 
> Why? An additional endpoint could report the total number of payments as well.

That could be done aswell, though I find it easier to use if its integrated into the endpoint.

**Comment by callebtc** [comment]
In that case we need the old endpoint to behave exactly the same as previously. This is the case in this PR, right?

File: lnbits/core/crud.py (hunk 2 above)
Comment (jackstar12) [reply lnbits/core/crud.py:433]: I didn't want to duplicate all the code, the functionality is the same

File: lnbits/core/static/js/wallet.js [hunk 5]
```diff
@@ -685,9 +690,26 @@ new Vue({
...
+        limit: pagination.rowsPerPage,
+        offset: (pagination.page - 1) * pagination.rowsPerPage,
+        sortby: pagination.sortBy,
+        direction: pagination.descending ? 'desc' : 'asc'
+      }
+      if (this.paymentsTable.filter) {
+        query.search = this.paymentsTable.filter
+      }
+      return LNbits.api.getPayments(this.g.wallet, query).then(response => {
+        this.paymentsTable.loading = false
```
Comment (motorina0) [inline lnbits/core/static/js/wallet.js:710]: `loading` should be set to false on error also

File: lnbits/core/crud.py [hunk 6]
```diff
@@ -395,21 +395,54 @@ async def get_payments(
         clause.append("checking_id NOT LIKE 'temp_%'")
         clause.append("checking_id NOT LIKE 'internal_%'")
 
+    return await (conn or db).fetch_page(
```
Comment (motorina0) [inline lnbits/core/crud.py:398]: - where is `fetch_page` defined?
   - I can find for example `fetchall` in `sqlalchemy_aio.base.AsyncResultProxy`
- also please note the `rewrite_query` calls from `lnbits/db.py/Connection`

File: lnbits/core/views/api.py [hunk 7]
```diff
@@ -34,7 +34,7 @@
     stop_extension_background_work,
 )
 from lnbits.core.models import Payment, User, Wallet
-from lnbits.db import Filters
+from lnbits.db import Filters, Page
```
Comment (motorina0) [inline lnbits/core/views/api.py:37]: where is `Page` defined? 
I could not find it.

File: lnbits/core/crud.py [hunk 8]
```diff
@@ -7,7 +7,7 @@
 import shortuuid
 
 from lnbits import bolt11
-from lnbits.db import COCKROACH, POSTGRES, Connection, Filters
+from lnbits.db import COCKROACH, POSTGRES, Connection, Filters, Page
```
Comment (motorina0) [inline lnbits/core/crud.py:10]: where is `Page` defined? 
I could not find it.

Comment (dni) [reply lnbits/core/crud.py:10]: https://github.com/lnbits/lnbits/pull/1613/files#diff-5855a9aec8f89dbcb613cd599d551e090dc55f68b23b20b7c2473261db814dadR111

its inside db.py

File: lnbits/core/crud.py (hunk 6 above)
Comment (dni) [reply lnbits/core/crud.py:398]: its defined int db.py aswell, its a fetch_page is a method of Database class.

File: lnbits/core/crud.py (hunk 8 above)
Comment (motorina0) [reply lnbits/core/crud.py:10]: ah, right
db.py content was hidden in the web view

**Review by dni — APPROVED** [review]

File: lnbits/db.py [hunk 9]
```diff
@@ -215,130 +428,21 @@ async def fetchone(self, query: str, values: tuple = ()):
             await result.close()
             return row
 
+    async def fetch_page(
+        self,
+        query: str,
+        where: Optional[list[str]] = None,
```
Comment (motorina0) [inline lnbits/db.py:434]: in all other places `List[str]` is used instead of `list[str]`

**Comment by talvasconcelos** [comment]
LGTM!

Closes #1182, closes #1519, closes #1620, closes #1652, closes lnbits/lndhub#3, closes lnbits/lndhub#4, closes lnbits/lndhub#5, closes lnbits/lndhub#6, closes lnbits/lndhub#7

**Comment by jackstar12** [comment]
> LGTM!
> 
> Closes #1182, closes #1519, closes #1620, closes #1652, closes [lnbits/lndhub#3](https://github.com/lnbits/lndhub/issues/3), closes [lnbits/lndhub#4](https://github.com/lnbits/lndhub/issues/4), closes [lnbits/lndhub#5](https://github.com/lnbits/lndhub/issues/5), closes [lnbits/lndhub#6](https://github.com/lnbits/lndhub/issues/6), closes [lnbits/lndhub#7](https://github.com/lnbits/lndhub/issues/7)

I'm not sure it closes all of these, but only https://github.com/lnbits/lnbits/issues/1620, https://github.com/lnbits/lnbits/issues/1652 while https://github.com/lnbits/lndhub/issues/7 could still be done but is not necessary anymore

**Review by dni — APPROVED** [review]

**Review by dni — APPROVED** [review]

**Review by callebtc — APPROVED** [review]
LGTM. Excellent work @jackstar12 you rock pretty hard.

**Review by callebtc — APPROVED** [review]
LGTM. Excellent work @jackstar12 you rock pretty hard.

**Pull request merged**